File: 61. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if head is None or head.next is None:
            return head
        if k == 0:
            return head
        else:
            temp = head
            count = 1
            while temp.next:
                temp = temp.next
                count += 1
            k = k % count
            if k == 0:
                return head
            curr = head
            prev = None
            for i in range(count - k):
                prev = curr
                curr = curr.next
            prev.next = None
            temp.next = head
            return curr

File: 61. Rotate List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }
        ListNode node = head;
        int length = 1;
        while (node.next != null) {
            node = node.next;
            length++;
        }
        k = k % length;
        ListNode prev = null;
        ListNode newhead = head;
        for (int i = 0; i < length-k;i++) {
            prev = newhead;
            newhead = newhead.next;
        }
        if (newhead == null) {
            return head;
        } else {
            prev.next = null;
            node.next = head;
            return newhead;
        }
    }
}